File: apps/api/app/services/agentmail.py
from __future__ import annotations

import asyncio
import html
import re

import markdown as markdown_lib
from agentmail import AgentMail
from svix.webhooks import Webhook

from app.config import settings


def verify_webhook(payload: bytes, headers: dict[str, str]) -> bool:
    if not settings.agentmail_webhook_secret:
        return settings.demo_mode
    try:
        Webhook(settings.agentmail_webhook_secret).verify(payload, headers)
    except Exception:
        return False
    return True


def _to_html(text: str) -> str:
    # Escape first so any raw HTML/script in Mike's output or a customer's message
    # cannot be injected into the outgoing email; markdown syntax still renders.
    body = markdown_lib.markdown(html.escape(text), extensions=["extra", "sane_lists"])
    return (
        '<div style="font-family: Arial, Helvetica, sans-serif; font-size: 14px; '
        'line-height: 1.6; color: #1a1a1a;">'
        f"{body}"
        "</div>"
    )


def _to_plain(text: str) -> str:
    text = re.sub(r"\*\*(.+?)\*\*", r"\1", text)  # bold
    text = re.sub(r"__(.+?)__", r"\1", text)
    text = re.sub(r"(?<![\w*])\*(?!\s)(.+?)(?<!\s)\*(?![\w*])", r"\1", text)  # italic
    text = re.sub(r"^\s{0,3}#{1,6}\s*", "", text, flags=re.MULTILINE)  # headings
    text = re.sub(r"\[([^\]]+)\]\((https?://[^)]+)\)", r"\1 (\2)", text)  # links
    return text.strip()


def _configured() -> bool:
    return not settings.demo_mode and bool(settings.agentmail_api_key) and bool(settings.agentmail_inbox_id)


async def reply_to_email(message_id: str, text: str, cc: list[str] | None = None) -> str | None:
    if not _configured():
        return None
    client = AgentMail(api_key=settings.agentmail_api_key)
    kwargs: dict = {
        "inbox_id": settings.agentmail_inbox_id,
        "message_id": message_id,
        "text": _to_plain(text),
        "html": _to_html(text),
    }
    if cc:
        kwargs["cc"] = cc
    reply = await asyncio.to_thread(lambda: client.inboxes.messages.reply(**kwargs))
    return getattr(reply, "message_id", None)


async def send_email(to: str, subject: str, text: str) -> str | None:
    if not _configured():
        return None
    client = AgentMail(api_key=settings.agentmail_api_key)
    sent = await asyncio.to_thread(
        lambda: client.inboxes.messages.send(
            inbox_id=settings.agentmail_inbox_id,
            to=[to],
            subject=subject,
            text=_to_plain(text),
            html=_to_html(text),
        )
    )
    return getattr(sent, "message_id", None)

File: lib/agentmail.ts
import { marked } from "marked";
import { Webhook } from "svix";
import { isDeployedEnvironment } from "@/lib/env";

function configured() {
  const demo = (process.env.DEMO_MODE || "").toLowerCase() === "true";
  return !demo && Boolean(process.env.AGENTMAIL_API_KEY) && Boolean(process.env.AGENTMAIL_INBOX_ID);
}

export function verifyWebhook(payload: Buffer, headers: Headers) {
  const secret = process.env.AGENTMAIL_WEBHOOK_SECRET;
  if (!secret) return !isDeployedEnvironment();
  try {
    const webhook = new Webhook(secret);
    webhook.verify(payload, {
      "svix-id": headers.get("svix-id") || "",
      "svix-timestamp": headers.get("svix-timestamp") || "",
      "svix-signature": headers.get("svix-signature") || "",
    });
    return true;
  } catch {
    return false;
  }
}

function toHtml(text: string) {
  const escaped = text
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
  const body = marked.parse(escaped, { async: false }) as string;
  return `<div style="font-family: Arial, Helvetica, sans-serif; font-size: 14px; line-height: 1.6; color: #1a1a1a;">${body}</div>`;
}

function toPlain(text: string) {
  return text
    .replace(/\*\*(.+?)\*\*/g, "$1")
    .replace(/__(.+?)__/g, "$1")
    .replace(/(?<![\w*])\*(?!\s)(.+?)(?<!\s)\*(?![\w*])/g, "$1")
    .replace(/^\s{0,3}#{1,6}\s*/gm, "")
    .replace(/\[([^\]]+)\]\((https?:\/\/[^)]+)\)/g, "$1 ($2)")
    .trim();
}

async function agentmailFetch(path: string, init: RequestInit) {
  const res = await fetch(`https://api.agentmail.to${path}`, {
    ...init,
    headers: {
      Authorization: `Bearer ${process.env.AGENTMAIL_API_KEY}`,
      "Content-Type": "application/json",
      ...(init.headers || {}),
    },
  });
  if (!res.ok) {
    throw new Error(`AgentMail request failed: ${res.status} ${await res.text()}`);
  }
  return res.json() as Promise<{ message_id?: string; id?: string }>;
}

export async function replyToEmail(messageId: string, text: string, cc?: string[] | null) {
  if (!configured()) return null;
  const inboxId = process.env.AGENTMAIL_INBOX_ID as string;
  const payload: Record<string, unknown> = {
    text: toPlain(text),
    html: toHtml(text),
  };
  if (cc?.length) payload.cc = cc;
  const reply = await agentmailFetch(`/v0/inboxes/${inboxId}/messages/${messageId}/reply`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
  return reply.message_id || reply.id || null;
}

export async function sendEmail(to: string, subject: string, text: string) {
  if (!configured()) return null;
  const inboxId = process.env.AGENTMAIL_INBOX_ID as string;
  const sent = await agentmailFetch(`/v0/inboxes/${inboxId}/messages`, {
    method: "POST",
    body: JSON.stringify({
      to: [to],
      subject,
      text: toPlain(text),
      html: toHtml(text),
    }),
  });
  return sent.message_id || sent.id || null;
}

export function agentmailOrgId() {
  const configuredOrg =
    process.env.AGENTMAIL_ORG_ID || process.env.MIKE_WIDGET_ORG_ID || process.env.MIKE_DEV_ORG_ID;
  if (configuredOrg) return configuredOrg;
  if (isDeployedEnvironment()) return null;
  return "dev-org";
}
